Sprint 2: Incr4 - Incr8

zueve opened the pull request.

zueve committed:
incr4: add tests

zueve committed:
Incr4: small refactoring

zueve committed:
Incr4: refactoring

zueve committed:
incr5: support config from env

zueve committed:
incr5: rename config/config -> config/context

zueve committed:
incr5: rename config/config -> config/context

zueve committed:
incr6: support file storage

zueve committed:
incr6: few fixes

zueve committed:
incr6: fix test in pipeline

zueve committed:
incr7: support cmd flags

zueve committed:
fix naming

zueve committed:
incr8: add compression

zueve committed:
incr8: fix reader

zueve committed:
incr8: support "application/x-gzip:

zueve committed:
incr8: support "application/x-gzip:

iTiky commented on internal/config/context.go:
```diff
@@ -0,0 +1,61 @@
+
+type ContextOption func(*Context)
+
+type Context struct {
```
Это очень плохое имя для структуры. В Go Context это строго `context.Context`.

zueve replied:
Согласен, но именно так описано в статье, на которую ты мне указал в прошлый раз https://medium.com/star-gazers/go-options-pattern-da49185a2526

iTiky commented on internal/config/context.go:
```diff
@@ -0,0 +1,61 @@
+type ContextOption func(*Context)
+
+type Context struct {
+	ctx             context.Context
```
Контекст никогда не должен быть внутри какого-либо состояния. Есть исключения, но здесь - это не оправдано (это точно не часть конфига).

zueve replied:
Тоже самое https://medium.com/star-gazers/go-options-pattern-da49185a2526

iTiky commented on internal/config/models.go:
```diff
@@ -0,0 +1,7 @@
+package config
```
Имя файла `models.go` сбивает с толку. Это не модели.

iTiky commented on internal/config/models.go:
```diff
@@ -0,0 +1,7 @@
+package config
+
+type Env struct {
+	ServerAddress   string `env:"SERVER_ADDRESS" envDefault:":8080"`
```
Почему эти структурные тэги нельзя было добавить в структуру Context? Лишнее дублирование сущностей.

zueve replied:
Прока можно, но представление в env может отличаться от других, и когда-нибудь их придется разделять

iTiky commented on internal/config/context.go:
```diff
@@ -0,0 +1,61 @@
+	return &uCtx
+}
+
+func NewContextFromEnvAndCMD() *Context {
```
Возврат указателя ничем не оправдан. Конфиг должен быть иммутабелен.

zueve replied:
Тоже самое https://medium.com/star-gazers/go-options-pattern-da49185a2526

iTiky commented on internal/config/context.go:
```diff
@@ -0,0 +1,61 @@
+	var envronment Env
+	err := env.Parse(&envronment)
+	if err != nil {
+		panic(err)
```
Решение о панике должен принимать только главный тред. В данном случае необходимо вернуть ошибку.

iTiky commented on internal/config/context.go:
```diff
@@ -0,0 +1,61 @@
+	FileStoragePath string
+}
+
+func NewContext(opts ...ContextOption) *Context {
```
Здорово, конечно, что ты применил этот паттерн, но здесь он лишний. Все поля этой структуры - публичны, их просто задать. Опции-сеттеры тут лишние.

iTiky commented on internal/config/context.go:
```diff
@@ -0,0 +1,61 @@
+	if err != nil {
+		panic(err)
+	}
+	baseURL := flag.String("b", envronment.BaseURL, "a string")
```
В данном варианте что имеет больший приоритет, ENV или flag? Флаги - это конфигурация CLI приложения. ENV-ы - это переопределение дефолтных значений. Иначе можно представить так: флаги - это Docker entrypoint, ENV-ы - это переменные Docker-а для настройки "под себя" и передачи секретов.

zueve replied:
тут не согласен. Флаги первичнее! Когда пишешь команду add -s 123 - ожидаешь что s всегда будет 123 независимо от параметров среду, которые вообще могут быть определены глобально, и даже иметь конфликты с приложением

iTiky commented on cmd/shortener/main.go:
```diff
@@ -14,12 +14,11 @@ import (
+	ctx := config.NewContextFromEnvAndCMD()
+	persistentStorage := storage.NewFileStorage(ctx.FileStoragePath)
+	defer persistentStorage.Close()
+	storageVar := storage.New(persistentStorage)
```
Читая строку `server.New(ctx, serviceVar)`, я понимаю что в функцию передают `context.Context`, но у тебя этот стандартный Go-паттерн ломается.

zueve replied:
Тоже самое https://medium.com/star-gazers/go-options-pattern-da49185a2526

iTiky commented on internal/server/server.go:
```diff
@@ -45,32 +48,36 @@ func (s *Server) Shutdown(ctx context.Context) error {
+	case "application/x-gzip":
+		urlBytes, err := io.ReadAll(r.Body)
+		if err != nil {
+			s.error(w, http.StatusInternalServerError, "invalid parse body")
```
"invalid parse body" это не очень по-русски =) Я бы ещё добавил сюда саму ошибку `err`

zueve replied:
A я бы не стал прокидывать наружу неизвестную информацию. В логах ок

iTiky commented on internal/server/server.go:
```diff
@@ -83,9 +90,47 @@ func (s *Server) redirect(w http.ResponseWriter, r *http.Request) {
+			s.error(w, http.StatusInternalServerError, "invalid parse body")
+			return
+		}
+		err = json.Unmarshal(dataBytes, &redirect)
```
Уже писал об этом ранее. Лучше придерживаться Go-стиля:
```go
if err := json.Unmarshal(...); err != nil {}
```

zueve replied:
Если честно, не нашел где ты на это указывал. В коде я такое сокращение встречал, но оно мне как-то дико не нравится.

iTiky commented on internal/server/server_test.go:
```diff
@@ -15,16 +17,21 @@ import (
 func TestServer_createRedirect(t *testing.T) {
-	var storageTest = storage.New()
-	var serviceTest = services.New(storageTest)
+	os.Remove(filename)
```
Не во всех ОС есть /tmp. Можно воспользоваться stdlib-ой:
https://yourbasic.org/golang/temporary-file-directory/

iTiky commented on internal/server/server_test.go:
```diff
@@ -153,4 +162,68 @@ func TestServer_redirect(t *testing.T) {
+			method:      http.MethodPost,
+			contentType: "application/json",
+			code:        201,
+			data:        "{\"url\": \"http://example.com\"}",
```
Это не читабельно. Лучше прям в этой фукнции объявить тип и получать его JSON-репрезентацию:
```go
type request struct {
    Field string `json:"field"`
}

tests := []struct {
    data request
}{
    {
        data: request {
            Field: "bla bla"
        }
    }
}
```

iTiky commented on internal/server/server_test.go:
```diff
@@ -153,4 +162,68 @@ func TestServer_redirect(t *testing.T) {
+			h.ServeHTTP(w, request)
+			res := w.Result()
+			if res.StatusCode != tt.code {
+				t.Errorf("Expected status code %d, got %d", tt.code, w.Code)
```
Как я писал ранее, лучше изучать testify библиотеку, с ней тесты станут более читаемы и поддерживаемы.
Эта строка будет выглядеть так:
```go
require.Equal(t, tt.code, res.StatusCode)
```

zueve replied:
Да как-то нет времени изучать...

iTiky commented on internal/storage/storage.go:
```diff
@@ -6,16 +6,27 @@ import (
+func New(persistent PersistentStorage) *Storage {
+	data, err := persistent.Load()
+	if err != nil {
+		panic(err)
```
Необходимо вернуть ошибку, но не паниковать.

iTiky commented on internal/storage/storage.go:
```diff
@@ -6,16 +6,27 @@ import (
 	"sync"
 )
 
+type PersistentStorage interface {
```
Имя сбивает с толку. Мне показалось, что ты объявил интерфейс, а ниже - его имплементация. Если ожидаемый контракт для зависимости, лучше его именовать как `PersistentStorageExpected`.

iTiky commented on internal/storage/file.go:
```diff
@@ -0,0 +1,81 @@
+		}
+	}
+	if err := scanner.Err(); err != nil {
+		log.Fatal(err)
```
Почему не возврат ошибки, а падение? Кроме того, в одном месте у тебя Fatal, в другом - паника. В обоих случаях должен быть возврат ошибки.

iTiky reviewed (changes requested):
Нужны доработки + не учтены все мои комментарии из прошлого PR-а.

zueve committed:
incr8: fix options

zueve committed:
incr8: fixes after code review

incr8: fix forward errors

incr8: rename PersistentStorageExpected

incr8: fix forward error

incr8: fix temt file

incr8: add untracked files

incr8: fix cm flags

incr8: fixes after 1-2-1

incr8: fix tests

iTiky commented on internal/server/server_test.go:
```diff
@@ -12,26 +13,68 @@ import (
+	persistentStorage, _ := storage.NewFileStorage(file.Name())
+	storageTest, err := storage.New(persistentStorage)
+	if err != nil {
+		return TestServer{}, err
```
Testify: `require.NoError(t, err)`
С помощью этой либы будет куда читабельнее.

iTiky reviewed (approved):
✔️
